targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F439xI/TARGET_MODULE_UBLOX_ODIN_W2/sdk/ublox-odin-w2-drivers/OdinWiFiInterface.cpp

```diff
@@ -0,0 +1,1854 @@
+    const char          *passwd;
+    uint8_t             channel;
+    nsapi_security_t    security;
+    unsigned int        timeout;
+};
+
+struct user_scan_s {
+    WiFiAccessPoint *ap_list;
+    nsapi_size_t    ap_list_size;
+};
+
+struct user_scan_rsp_s {
+    WiFiAccessPoint         *ap_list;
+    nsapi_size_or_error_t   found_aps;
+};
+
+struct user_response_s {
+    nsapi_error_t error;
+};
+
+struct user_ap_start_s {
+    const char          *ssid;
+    const char          *passwd;
+    uint8_t             channel;
+    nsapi_security_t    security;
+    uint16_t            beacon_interval;
+};
+
+struct wlan_status_started_s {
+    cbWLAN_StatusStartedInfo info;
+};
+
+struct wlan_status_connected_s {
+    cbWLAN_StatusConnectedInfo info;
+};
+
+struct wlan_status_connection_failure_s {
+    cbWLAN_StatusDisconnectedInfo info;
+};
+
+struct odin_wifi_msg_s {
+    unsigned int                                type;
+
+    union data_t {
+        struct user_connect_s                   user_connect;
+        struct user_response_s                  user_response;
+        struct user_scan_s                      user_scan;
+        struct user_scan_rsp_s                  user_scan_rsp;
+        struct user_ap_start_s                  user_ap_start;
+        struct wlan_status_started_s            wlan_status_started;
+        struct wlan_status_connected_s          wlan_status_connected;
+        struct wlan_status_connection_failure_s wlan_status_connection_failure;
+    } data;
+};
+
+static void generateWEPKeys(const char *passphrase, cbWLAN_WEPKey keys[4]);
+
+static void set_wpa_rsn_cipher(
+        nsapi_security_t    security,
+        cbWLAN_CipherSuite  &wpa_ciphers,
+        cbWLAN_CipherSuite  &rsn_ciphers);
+
+static bool is_valid_AP_channel(cbWLAN_Channel channel);
+
+// Friend declared C-functions that calls corresponding wi-fi object member function
+struct wlan_callb_s {
+
+    static void scan_indication_callback(void *callb_context, cbWLAN_ScanIndicationInfo *scan_info, cb_boolean is_last_result)
+    {
+        MBED_ASSERT(callb_context != NULL);
+
+        OdinWiFiInterface *wifi = (OdinWiFiInterface*)callb_context;
+
+        wifi->wlan_scan_indication(scan_info, is_last_result);
+    };
+
+    static void status_indication_callback(void *callb_context, cbWLAN_StatusIndicationInfo status, void *data)
+    {
+        MBED_ASSERT(callb_context != NULL);
+
+        OdinWiFiInterface *wifi = (OdinWiFiInterface*)callb_context;
+
+        wifi->wlan_status_indication(status, data);
+    };
+
+    static void odin_thread_fcn(OdinWiFiInterface *wifi)
+    {
+        wifi->handle_in_msg();
+    };
+};
+
+// Static declared class variables
+bool OdinWiFiInterface::_wlan_initialized = false;
+int32_t OdinWiFiInterface::_target_id = cbMAIN_TARGET_INVALID_ID;
+
+OdinWiFiInterface::OdinWiFiInterface(): 
+    _thread(osPriorityNormal, 4096),
+    EMACInterface(OdinWiFiEMAC::get_instance(), OnboardNetworkStack::get_default_instance())
```

Comment:
You must support arbitrary network stacks being specified, and you should do it via the same mechanism as `EthernetInterface` - having a default parameter in your constructor:

      OdinWiFiInterface(OnboardNetworkStack &stack = OnboardNetworkStack::get_default_instance()) : 
          EMACInterface(OdinWiFiEMAC::get_instance(), stack)

      OdinWiFiInterface(bool debug, OnboardNetworkStack &stack = OnboardNetworkStack::get_default_instance())

I think I let this pass in the original binary version, but now would be a good time to fix it. Ultimately we'd like to be able to use this as a Wi-fi Thread/6LoWPAN border router, and that would be a prerequisite.